### PR TITLE
[annotation] Update DraggableAnnotationController for multiple draggable managers (#863, #883)

### PR DESCRIPTION
**Pull request by Ph0tonic**
This is a proposal for fixing #863 and #883.

They are kind of linked together and I foud the solution of the second one during my work so I added it in this PR.

I know that it's not perfect but at least is a simple solution which allow multiple manager to be draggable.

@LukasPaczos and @langsmith I would be happy to have your feedback. 👍

### CODE REVIEW

**Comment by LukasPaczos** [comment]
Hey @Ph0tonic, thanks for your contribution again!

Unfortunately, I don't think this is an approach that we'd like to take. Having multiple detectors running is going to get inefficient.

To resolve the issue, I'd rather explore the option of having one, public instance of the draggable controller into which we can inject multiple annotation managers, instead of each annotation manager creating its own draggable controller.

**Comment by Ph0tonic** [comment]
Ok, I see, I will try to turn draggable controller into a singleton such has to minimize the number of detectors.

**Comment by Ph0tonic** [comment]
@LukasPaczos Here is a completely new implementation which turns DraggableController into a singleton. Would this be closer to an approach you would take ?

**Comment by Ph0tonic** [comment]
I had a look at the `make generate-annotation-code` and the linked scripts and I think that some of the `.ejs` needs to be updated but as these are annotated with:
```java
// This file is generated.
```

I let those like they currently are. Please let me known if I can do anything else @langsmith and @LukasPaczos. Thanks

**Comment by Ph0tonic** [comment]
The limitations resolved by this PR would really help me for one of my side project. If we arrive to a solution, do you think we might be able to do a release for those next weeks ?
@LukasPaczos and @langsmith Thank you very much for your time.

**Comment by kikothemaster** [comment]
When can we expect this fix in production?

**Comment by Ph0tonic** [comment]
Any updates about this PR ? @LukasPaczos and @langsmith

File: plugin-annotation/src/main/java/com/mapbox/mapboxsdk/plugins/annotation/DraggableAnnotationController.java [hunk 1]
```diff
@@ -16,17 +16,32 @@
...
+final class DraggableAnnotationController <T extends Annotation, D extends OnAnnotationDragListener<T>> {
+
+  private static DraggableAnnotationController INSTANCE = null;
+
+  public static synchronized DraggableAnnotationController getInstance(MapView mapView, MapboxMap mapboxMap) {
+    if(INSTANCE == null){
+      INSTANCE = new DraggableAnnotationController(mapView, mapboxMap);
+    }
+    return INSTANCE;
+  }
```
Comment (LukasPaczos) [inline plugin-annotation/src/main/java/com/mapbox/mapboxsdk/plugins/annotation/DraggableAnnotationController.java:31]: An issue with this singleton approach might occur when the map is recreated, for example in a recycler view, or when transitioning between fragments/activites.

We might need to do something like this, to additionally ensure that we're using the correct instance:
```suggestion
  public static synchronized DraggableAnnotationController getInstance(MapView mapView, MapboxMap mapboxMap) {
    if(INSTANCE == null || this.mapView != mapview || this.mapboxMap != mapboxMap){
      INSTANCE = new DraggableAnnotationController(mapView, mapboxMap);
    }
    return INSTANCE;
  }
```

The strong references to `mapView` and `mapboxMap` should be dropped when the last `removeAnnotationManager` is called and the instance is nullified so we shouldn't leak. It'd be great to test this scenario though.

Comment (Ph0tonic) [reply plugin-annotation/src/main/java/com/mapbox/mapboxsdk/plugins/annotation/DraggableAnnotationController.java]: You're right I did not think about that 👍. I updated the code to reflect what you said. Regarding the testing of the last part of your remark, I'm not sure how I can do it.

**Comment by LukasPaczos** [comment]
Amazing work here, thanks @Ph0tonic! Let's continue in https://github.com/mapbox/mapbox-plugins-android/pull/1147.

**Comment by Ph0tonic** [comment]
> 
> 
> Amazing work here, thanks @Ph0tonic! Let's continue in #1147.

You are welcome

**Pull request closed without merging**